src/cli/util.js

```diff
@@ -835,11 +844,13 @@ function updateContextOptions(context, plugins) {
 
   const apiDefaultOptions = supportOptions
     .filter(optionInfo => !optionInfo.deprecated)
-    .reduce(
-      (reduced, optionInfo) =>
-        Object.assign(reduced, { [optionInfo.name]: optionInfo.default }),
-      Object.assign({}, optionsModule.hiddenDefaults)
-    );
+    .reduce((reduced, optionInfo) => {
+      // default are applied in main/options.js when pluginDefaults are present
+      if (Object.keys(optionInfo.pluginDefaults).length) {
+        return reduced;
+      }
+      return Object.assign(reduced, { [optionInfo.name]: optionInfo.default });
+    }, Object.assign({}, optionsModule.hiddenDefaults));
```

Comment:
I managed to get the tests passing by adding this workaround - but I'm not really happy with it. I could imagine that there's a case where this breaks, and even if there isn't it doesn't look very clean to me.

I'd be glad if someone could explain the bigger picture on the way defaults are handled right now: To me, the two-step process of applying defaults seems confusing:

1. CLI defaults are applied in `cli/util.js`
https://github.com/prettier/prettier/blob/9ee113cd2e66dad5567ce03170d1c6c6dff1a310/src/cli/util.js#L226
2. Defaults are applied if an option is `null` in `main/options.js` 
https://github.com/prettier/prettier/blob/9ee113cd2e66dad5567ce03170d1c6c6dff1a310/src/main/options.js#L58

I saw that there are some use cases where the core options are still `null` in step two, but I'm not sure what those use cases are, and consequently how the proper process of applying defaults should look like in the future.

Reply:
--> #4045